packages/api/src/utils/sqlObfuscator.ts

```diff
@@ -0,0 +1,47 @@
+import { spawn } from 'child_process';
+
+let subprocess: any;
+
+const getChild = () => {
+  if (subprocess) {
+    return subprocess;
+  }
+
+  const arch = process.arch;
+
+  subprocess = spawn(`src/gobin/sql_obfuscator_${arch}`, [], {
+    stdio: ['pipe', 'pipe', 'pipe', 'ipc'],
+  });
+
+  process.on('SIGINT', () => {
+    subprocess.kill('SIGINT');
+  });
+
+  return subprocess;
+};
+
+export const sqlObfuscator = async (sql: string): Promise<string> => {
+  subprocess = getChild();
+  const strippedSql = sql.replace(/(\r\n|\n|\r)/gm, ' ');
+  subprocess.stdin.write(`${strippedSql}\n`);
+
+  return new Promise((resolve, reject) => {
+    const removeListeners = (subprocess: any) => {
+      subprocess.stdout.removeAllListeners();
+      subprocess.stderr.removeAllListeners();
+      subprocess.removeAllListeners();
+    };
+    const errorOutput = (data: any) => {
+      removeListeners(subprocess);
+      reject(data.toString());
+    };
+    const dataRecieved = (data: any) => {
+      removeListeners(subprocess);
+      resolve(data.toString());
+    };
+    subprocess.stdout.on('data', dataRecieved);
```

Comment:
Trying to understand this. how do we know if all stdout are streamed back already ? what if its more than a buffer ? I wonder if we want to aggregate the data at 'close' event and resolve the promise from there

Reply:
We can't use the `close` event because it would EOF the stdout from the process. I should expand it to check for a `\n` at the end, that's the terminator. Basically, the subprocess reads until it finds a newline, then emits the obfuscated sql and a newline, so if we see a newline then the buffer has been flushed. I will experiment with the `end` event and see if we can get better confidence